obspy/clients/httpproxy.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+"""
+Establish a socket connection through an HTTP proxy.
+Author: Fredrik Østrem <frx.apps@gmail.com>
+License:
+  This code can be used, modified and distributed freely, as long as it is this
+  note containing the original author, the source and this license, is put
+  along with the source code.
+
+J. MacCarthy, modified from https://gist.github.com/frxstrem/4487802
+
+"""
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+from future import standard_library
+from future.utils import native_str
+
+from base64 import b64encode
+import socket
+with standard_library.hooks():
+    import urllib
+    import urlparse
+
+
+def get_proxy_tuple():
+    """
+    Return system http proxy as a urlparse.urlparse tuple or () if unset.
+    """
+    proxydict = urllib.getproxies()
+    proxystr = proxydict.get('http') or proxydict.get('https') or ''
+    if proxystr:
+        proxy = urlparse.urlparse(proxystr)
+    else:
+        proxy = ()
+
+    return proxy
+
+
+def valid_address(addr):
+    """ Verify that an IP/port tuple is valid """
+    is_valid = (isinstance(addr, (list, tuple)) and
+                len(addr) == 2 and
+                isinstance(addr[0], (str, native_str)) and
+                isinstance(addr[1], int))
+    return is_valid
+
+
+def http_proxy_connect(address, proxy, auth=None, headers={}):
+    """
+    Establish a socket connection through an HTTP proxy.
+
+    Arguments:
+      address (required)     = The address of the target
+      proxy (def: None)      = The address of the proxy server
+      auth (def: None)       = A tuple of the username and password used for
+                               authentication
+      headers (def: {})      = A set of headers that will be sent to the proxy
+
+    Returns:
+      A 3-tuple of the format:
+        (socket, status_code, headers)
+      Where `socket' is the socket object, `status_code` is the HTTP status
+      code that the server returned and `headers` is a dict of headers that the
+      server returned.
+
+    """
+    if not valid_address(address):
+        raise ValueError('Invalid target address')
+
+    if not valid_address(proxy):
+        raise ValueError('Invalid proxy address')
+
+    headers = {'host': address[0]}
+
+    if auth is not None:
+        if isinstance(auth, str):
+            headers['proxy-authorization'] = auth
+        elif auth and isinstance(auth, (tuple, list)) and len(auth) == 2:
+            proxy_authorization = 'Basic ' + b64encode('%s:%s' % auth)
```

Comment:
`b64encode` accepts and returns `bytes`, so you won't be able to add these or call it.